crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -2344,6 +2344,352 @@ pub fn local_equivalence(weyl: PyReadonlyArray1<f64>) -> PyResult<[f64; 3]> {
     Ok([g0_equiv + 0., g1_equiv + 0., g2_equiv + 0.])
 }
 
+/// invert 1q gate sequence
+fn invert_1q_gate(gate: (StandardGate, SmallVec<[f64; 3]>)) -> (StandardGate, SmallVec<[f64; 3]>) {
+    let gate_params = gate.1.into_iter().map(Param::Float).collect::<Vec<_>>();
+    let Some(inv_gate) = gate
+        .0
+        .inverse(&gate_params) else {panic!()};
+    let inv_gate_params = inv_gate
+        .1
+        .into_iter()
+        .map(|param| match param {
+            Param::Float(val) => val,
+            _ => panic!(),
+        })
+        .collect::<SmallVec<_>>();
+    (inv_gate.0, inv_gate_params)
+}
+
+/// invert 2q gate sequence
+fn invert_2q_gate(
+    gate: (Option<StandardGate>, SmallVec<[f64; 3]>, SmallVec<[u8; 2]>),
+) -> (StandardGate, SmallVec<[f64; 3]>, SmallVec<[u8; 2]>) {
+    let Some(inv_gate) = gate
+                    .0
+                    .unwrap()
+                    .inverse(&gate.1.into_iter().map(Param::Float).collect::<Vec<_>>()) else {panic!()};
+    let inv_gate_params = inv_gate
+        .1
+        .into_iter()
+        .map(|param| match param {
+            Param::Float(val) => val,
+            _ => panic!(),
+        })
+        .collect::<SmallVec<_>>();
+    (inv_gate.0, inv_gate_params, gate.2)
+}
+
+#[pyclass(module = "qiskit._accelerate.two_qubit_decompose", subclass)]
+pub struct TwoQubitControlledUDecomposer {
+    #[pyo3(get)]
+    rxx_equivalent_gate: StandardGate,
+    #[pyo3(get)]
+    scale: f64,
+}
+
+const DEFAULT_ATOL: f64 = 1e-12;
+
+///  Decompose two-qubit unitary in terms of a desired
+///  :math:`U \sim U_d(\alpha, 0, 0) \sim \text{Ctrl-U}`
+///  gate that is locally equivalent to an :class:`.RXXGate`.
+impl TwoQubitControlledUDecomposer {
+    ///  Takes an angle and returns the circuit equivalent to an RXXGate with the
+    ///  RXX equivalent gate as the two-qubit unitary.
+    ///  Args:
+    ///      angle: Rotation angle (in this case one of the Weyl parameters a, b, or c)
+    ///  Returns:
+    ///      Circuit: Circuit equivalent to an RXXGate.
+    ///  Raises:
+    ///      QiskitError: If the circuit is not equivalent to an RXXGate.
+    fn to_rxx_gate(&self, angle: f64) -> PyResult<TwoQubitGateSequence> {
+        let mat = self
+            .rxx_equivalent_gate
+            .matrix(&[Param::Float(self.scale * angle)])
+            .unwrap();
+        let decomposer_inv =
+            TwoQubitWeylDecomposition::new_inner(mat.view(), Some(DEFAULT_FIDELITY), None)?;
+
+        let euler_basis = EulerBasis::ZYZ;
+        let mut target_1q_basis_list = EulerBasisSet::new();
+        target_1q_basis_list.add_basis(euler_basis);
+
+        let mut gates = Vec::new();
```

Comment:
I'd use `with_capacity()` here instead of `new()` so that we pre-allocate the maximum number of gates in the sequence. 


```suggestion
        let mut gates = Vec::with_capacity(13);
```
I got 13 from 3x gates for each of the 4x 1q components + the 2q gate.

Reply:
Done in [1ceb94e](https://github.com/Qiskit/qiskit/pull/13139/commits/1ceb94e13e9f3abee67736b19017dc3063fdffb2)